lib/ohai/plugins/linux/network.rb

```diff
@@ -332,15 +341,22 @@ def favored_default_route(routes, iface, default_route, family)
         # using the source field when it's specified :
         # 1) in the default route
         # 2) in the route entry used to reach the default gateway
-        r[:src] and # it has a src field
-          iface[r[:dev]] and # the iface exists
-          iface[r[:dev]][:addresses].has_key? r[:src] and # the src ip is set on the node
-          iface[r[:dev]][:addresses][r[:src]][:scope].downcase != "link" and # this isn't a link level addresse
-          ( r[:destination] == "default" or
-            ( default_route[:via] and # the default route has a gateway
-              IPAddress(r[:destination]).include? IPAddress(default_route[:via]) # the route matches the gateway
-              )
+        r[:src] && # it has a src field
+          iface[r[:dev]] && # the iface exists
+          (
+            iface[r[:dev]][:addresses].nil? || # this int has no addresses
+            iface[r[:dev]][:addresses].values.all? { |addr| addr['family'] != family[:name] } || # this int has no ip
+            (
+              iface[r[:dev]][:addresses].has_key?(r[:src]) && # the src ip is set on the node
+              iface[r[:dev]][:addresses][r[:src]][:scope].downcase != "link" # this isn't a link level address
+            )
+          ) && (
+            r[:destination] == "default" ||
+            (
+              default_route[:via] && # the default route has a gateway
+              IPAddress(r[:destination]).include?(IPAddress(default_route[:via])) # the route matches the gateway
             )
+          )
```

Comment:
this logic is kind of horrible to read.  the comments are useful.

could we extract some helpers out of this though in order to make the logic more readable?

a trivial example:

```ruby
def iface_has_no_addresses?(iface)
  iface[iface][:addresses].nil?
end
```